src/net/ipv4.c

```diff
@@ -373,7 +373,9 @@ static int ipv4_tx ( struct io_buffer *iobuf,
 		ntohs ( iphdr->chksum ) );
 
 	/* Calculate link-layer destination address, if possible */
-	if ( ( ( next_hop.s_addr ^ INADDR_BROADCAST ) & ~netmask.s_addr ) == 0){
+	if ( ( netmask.s_addr != INADDR_NET_RFC3021 ) &&
+	     ( ( ( next_hop.s_addr ^ INADDR_BROADCAST )
+		 & ~netmask.s_addr ) == 0) ) {
```

Comment:
This looks as though it would also cause packets sent to `INADDR_BROADCAST` to be transmitted as unicasts.  Is this correct behaviour?  I couldn't immediately find anything in RFC3021 that specifies how explicit broadcasts to `INADDR_BROADCAST` (e.g. DHCP requests) are supposed to be handled.

Reply:
Ah, yes, good point. From what I can tell, this change should only affect directed broadcasts, link-local broadcasts (to `INADDR_BROADCAST`) should still be broadcasts on L2 as well.
I'll update the PR to implement this and will test if it still works as expected. The `if` expression might become quite ugly though, I should probably split it out into a `inaddr_is_broadcast` helper function that does:
```
if INADDR_BROADCAST: return true
if /31 netmask: return false
return (all host bits are 1)
```

Reply:
Thanks for looking into it.  Please avoid adding more than two x86 assembly instructions to the overall test if possible.  :slightly_smiling_face:

Reply:
New version is pushed.
I didn't find a good way to get the test down to two instructions, it's currently adding four in total (a `cmp` and `je` each for the two checks before the existing one). In theory it might be possible to avoid the first test since the same test is already performed before the route lookup, but that would require more nested logic (if 255.255.255.255, set `ll_dest` to the broadcast address, otherwise perform route lookup and check if the nexthop is a directed broadcast).
I think the performance impact should be negligible though, there are no extra memory accesses and the branch predictor should handle it fine since most packets will go to the same destination.